components/cloud/aws/src/s3.rs

```diff
@@ -0,0 +1,598 @@
+                secret_access_key: StringNonEmpty::required_field(
+                    input.secret_access_key,
+                    "secret_access_key",
+                )?,
+            }),
+        };
+        Ok(Config {
+            storage_class,
+            bucket: config_bucket,
+            sse: StringNonEmpty::opt(input.sse),
+            acl: StringNonEmpty::opt(input.acl),
+            access_key_pair,
+            force_path_style: input.force_path_style,
+            sse_kms_key_id: StringNonEmpty::opt(input.sse_kms_key_id),
+        })
+    }
+}
+
+impl BlobConfig for Config {
+    fn name(&self) -> &'static str {
+        &STORAGE_NAME
+    }
+
+    fn url(&self) -> io::Result<url::Url> {
+        self.bucket.url("s3").map_err(|s| {
+            io::Error::new(
+                io::ErrorKind::Other,
+                format!("error creating bucket url: {}", s),
+            )
+        })
+    }
+}
+
+#[derive(Clone)]
+pub struct S3Storage {
+    config: Config,
+    client: S3Client,
+}
+
+impl S3Storage {
+    pub fn from_input(input: InputConfig) -> io::Result<Self> {
+        Self::new(Config::from_input(input)?)
+    }
+
+    pub fn from_cloud_dynamic(cloud_dynamic: &CloudDynamic) -> io::Result<Self> {
+        Self::new(Config::from_cloud_dynamic(cloud_dynamic)?)
+    }
+
+    /// Create a new S3 storage for the given config.
+    pub fn new(config: Config) -> io::Result<S3Storage> {
+        Self::with_request_dispatcher(config, util::new_http_client()?)
+    }
+
+    fn new_creds_dispatcher<Creds, Dispatcher>(
+        config: Config,
+        dispatcher: Dispatcher,
+        credentials_provider: Creds,
+    ) -> io::Result<S3Storage>
+    where
+        Creds: ProvideAwsCredentials + Send + Sync + 'static,
+        Dispatcher: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        let bucket_region = none_to_empty(config.bucket.region.clone());
+        let bucket_endpoint = config.bucket.endpoint.clone();
+        let region = util::get_region(&bucket_region, &none_to_empty(bucket_endpoint))?;
+        let client = S3Client::new_with(dispatcher, credentials_provider, region);
+        Ok(S3Storage { config, client })
+    }
+
+    pub fn with_request_dispatcher<D>(config: Config, dispatcher: D) -> io::Result<S3Storage>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        // TODO: this should not be supported.
+        // It implies static AWS credentials.
+        if let Some(access_key_pair) = &config.access_key_pair {
+            let cred_provider = StaticProvider::new_minimal(
+                (*access_key_pair.access_key).to_owned(),
+                (*access_key_pair.secret_access_key).to_owned(),
+            );
+            Self::new_creds_dispatcher(config, dispatcher, cred_provider)
+        } else {
+            let cred_provider = util::CredentialsProvider::new()?;
+            Self::new_creds_dispatcher(config, dispatcher, cred_provider)
+        }
```

Comment:
i don't know why we need that TODO comment. static credentials provided by the client is needed for non-cloud clusters.